Fix: Hidden container resize issue (This partially addresses #2659 — handling redraw when the viewer is initialized in a hidden container.)

amudhanj26 opened the pull request:
Fix for hidden container rendering issue.
This pull request addresses a common issue where OpenSeadragon doesn't render images correctly if the viewer is placed inside a container that is hidden (for example, using display: none) when it's first created. When that container later becomes visible, the viewer doesn't automatically update or redraw, so the image remains blank or doesn't load properly.

What I did is as follows, 

1.Added a check that watches for when the viewer's container becomes visible.

2.When it does, the viewer is forced to resize and redraw itself based on the actual container size.

3.This makes sure the image appears correctly when toggling visibility, like when using tabs or collapsible panels.

These are needed because, many applications use hidden containers for things like tabs or popups. Without this fix, the viewer doesn’t display images properly unless the developer manually triggers a redraw. This change helps make OpenSeadragon work better out of the box in real-world interfaces.

I tested it by creating a simple HTML page where the viewer starts inside a hidden `<div>` that can be toggled visible. With the fix, the image now appears correctly every time the container is shown.

Let me know if you’d like any changes.

amudhanj26 commented:
Hi @iangilman,  I tested this locally by rebuilding the project and running the full test suite via http://localhost:8000/test/. The tests that fail — especially `Basic: FullScreen', seems unrelated to the code I changed. From what I can tell, they're due to fullscreen being blocked by the browser (likely because the test wasn't triggered by a trusted user interaction).

Let me know if you want me to adjust anything further!

iangilman reviewed (changes requested):
I actually can't get the bug you're fixing to happen. Can you share a repro? Here's what I tried: 

```
var viewer = OpenSeadragon({
    id: "contentDiv",
    prefixUrl: "../../build/openseadragon/images/",
    tileSources: "../data/testpattern.dzi",
});

setTimeout(() => {
    document.getElementById("contentDiv").style.display = "block";
}, 5000);
```

We already have a `ResizeObserver` which appears to be triggered properly when the viewer becomes visible. I assume you've found a circumstance where that doesn't work? 

BTW, you have tagged #2660 as the bug you're fixing, but that is a different issue. Can you please update for the correct issue?

I have not looked into the test failures, but I did restart the build and they still happen. My guess is they are from the `setTimeout` in `open` (causing things to happen after the viewer is shut down during a test). 

I appreciate the contribution! I just want to make sure we're doing the right thing.

amudhanj26 commented:
Hi @iangilman ,
Thanks so much for taking the time to review and provide detailed feedback, was really helpful.

I've now removed the setTimeout fallback and refactored the code to use IntersectionObserver with _this, consistent with the existing codebase style. I also corrected the PR title and linked it to the appropriate issue: #2659 (handling redraw when the viewer is initialised in a hidden container).

The Travis CI build passed, and only the GitHub Actions check is failing, please let me know if you'd like me to look into that.

Also let me know if anything else needs to be updated, happy to follow up!

iangilman commented on src/viewer.js:
```diff
@@ -267,25 +267,21 @@ $.Viewer = function( options ) {
-    const self = this;
-
-    if (typeof IntersectionObserver !== 'undefined') {
+    if (typeof window.IntersectionObserver !== 'undefined') {
```
`undefined` should not be in quotes here.

iangilman commented on src/viewer.js:
```diff
@@ -267,25 +267,21 @@ $.Viewer = function( options ) {
         }, { threshold: 0.1 });
 
-        observer.observe(this.container);
+        observer.observe(_this.container);
```
This doesn't need to be a `_this` since it's not inside the closure of the `forEach` function.

iangilman reviewed (changes requested):
Great, looking better! A couple minor notes.

I'd still love to know how to reproduce the issue. As I said, it seems to be working even without your patch as far as I can tell. I want to understand why the patch is needed.

Also, #2659 does not appear to be the correct issue either.

amudhanj26 commented:
Hi @iangilman, thanks a lot for reviewing and giving feedback, really appreciate you taking the time.

Just to be clear, I originally came across issue #2659 and was also reading through #2660, trying to understand both. While doing that, I noticed a bug where the OpenSeadragon viewer doesn’t display correctly if the container is hidden (display: none) when the page loads. That’s when I started experimenting and made this patch.

At that point, I wasn’t 100% sure which issue it was related to, and I got a bit confused while linking it. Sorry about the confusion caused. It is not directly solving #2659 or #2660, but rather this hidden-container case I found while testing. I understand now that I should’ve opened a separate issue for that. Let me know if I should do that. Apologies.

amudhanj26 commented:
Here’s how I was able to reproduce the issue this patch is meant to solve:

1. Create a <div> with style="display: none" and put the OpenSeadragon viewer inside it.

2. After the page has fully loaded, change the container’s style to display: block using JavaScript (for example with setTimeout or after a user action).

3. You will see the viewer does not load or render correctly, as the image appears blank or misaligned.

4. If you manually resize the browser window, it suddenly fixes itself and displays properly.

This is the situation I accidentally got caught up with, and this is why I added the IntersectionObserver to detect when the container becomes visible and trigger a resize/redraw automatically. Let me know if you'd prefer I open a new issue to track this behaviour separately. Happy to do that :)

iangilman commented:
Thank you for the explanations and for the example code! I've tried your test code on master (without the changes from this PR) and the viewer updates fine for me. When I refresh the page it is blank for two seconds and then the viewer shows up without me having to do anything. If I put a console log in the ResizeObserver in viewer.js, I see that it gets triggered when the element becomes visible.

I'm on Mac Chrome. I've tested it as well in Mac Firefox and Safari. It's working in all of them. What browser are you testing in? I'd like to get to the bottom of why it's working for me but not for you.
